java/src/main/java/org/rocksdb/RocksDB.java

```diff
@@ -60,36 +87,54 @@ public static synchronized void loadLibrary() {
    * @param paths a list of strings where each describes a directory
    *     of a library.
    */
-  public static synchronized void loadLibrary(final List<String> paths) {
-    for (CompressionType compressionType : CompressionType.values()) {
-      if (compressionType.equals(CompressionType.NO_COMPRESSION)) {
-        continue;
+  public static void loadLibrary(final List<String> paths) {
+    if (libraryLoaded.get() == LibraryState.LOADED) {
+      return;
+    }
+
+    if (libraryLoaded.compareAndSet(LibraryState.NOT_LOADED,
+        LibraryState.LOADING)) {
+      for (final CompressionType compressionType : CompressionType.values()) {
+        if (compressionType.equals(CompressionType.NO_COMPRESSION)) {
+          continue;
+        }
+        for (final String path : paths) {
+          try {
+            System.load(path + "/" + Environment.getSharedLibraryFileName(
+                compressionType.getLibraryName()));
+            break;
+          } catch (UnsatisfiedLinkError e) {
+            // since they are optional, we ignore loading fails.
+          }
+        }
       }
-      for (String path : paths) {
+      boolean success = false;
+      UnsatisfiedLinkError err = null;
+      for (final String path : paths) {
         try {
-          System.load(path + "/" + Environment.getSharedLibraryFileName(
-              compressionType.getLibraryName()));
+          System.load(path + "/" +
+              Environment.getJniLibraryFileName("rocksdbjni"));
+          success = true;
           break;
         } catch (UnsatisfiedLinkError e) {
-          // since they are optional, we ignore loading fails.
+          err = e;
         }
       }
+      if (!success) {
```

Comment:
Can we set the status to LibraryState.ERROR so that in case someone change the logic in the future, we won't have other thread goes into an infinite loop on line 131-136?